File: python/search/search_algos.py

class BinarySearch:
    def __init__(self):
        pass

    def recursive_search(self, sorted_list, item):
        """
        Searches to find if item is in the sorted list.
        Method calls _search_range() recursively.
        """
        low = 0
        high = len(sorted_list) - 1
        answer = self._search_range(sorted_list, high, low, item)
        return answer


    def _search_range(self, sorted_list, high, low, item):
        if low > high:
            return False
        mid = (high - low)//2 + low
        if item > sorted_list[mid]:
            low = mid + 1
        if item < sorted_list[mid]:
            high = mid - 1
        if item == sorted_list[mid]:
            return True
        return self._search_range(sorted_list, high, low, item)

    def iterative_search(self, sorted_list, item):
        """
        Searches to find if item is in the sorted list.
        Method runs iteratively.
        """
        high = len(sorted_list)-1
        low = 0
        while(True):
            if low > high:
                return False
            mid = (high-low)//2 + low
            if item > sorted_list[mid]:
                low = mid+1
            if item < sorted_list[mid]:
                high = mid-1
            if item == sorted_list[mid]:
                return True

# Naive SequentialSearch implemented for benchmark comparisons
class SequentialSearch:
    def __init__(self):
        pass

    def search(self, sorted_list, item):
        for i in sorted_list:
            if i == item:
                return True
        return False

File: cpp/search/src/search_algos.cpp
#include "search_algos.h"

typedef std::vector<int> vec;

bool BinarySearch::recursive_search(const vec& sorted_list, const int item){

    int low = 0;
    int high = sorted_list.size() - 1;
    bool answer = search_range(sorted_list, high, low, item);
    return answer;
}

bool BinarySearch::iterative_search(const vec& sorted_list, const int item){

    int low = 0;
    int high = sorted_list.size() - 1;
    while(true){
        if(low > high){
            return false;
        }
        int mid = (high - low)/2 + low;
        if(item > sorted_list[mid]){
            low = mid + 1;
        }
        if(item < sorted_list[mid]){
            high = mid - 1;
        }
        if(item == sorted_list[mid]){
            return true;
        }
    }
}

bool BinarySearch::search_range(const vec& sorted_list, int high, int low, const int item){

    if(low > high){
        return false;
    }
    int mid = (high - low)/2 + low;
    if(item > sorted_list[mid]){
        low = mid + 1;
    }
    if(item < sorted_list[mid]){
        high = mid - 1;
    }
    if(item == sorted_list[mid]){
        return true;
    }
    return search_range(sorted_list, high, low, item);
}

bool SequentialSearch::search(const vec& sorted_list, const int item){

    for (int i = 0; i < sorted_list.size(); i++){
        if(sorted_list[i] == item){
            return true;
        }
    }
    return false;
}